state/protocol/badger/snapshot.go

```diff
@@ -261,6 +249,68 @@ func (s *Snapshot) Commit() (flow.StateCommitment, error) {
 	return seal.FinalState, nil
 }
 
+func (s *Snapshot) LatestSeal() (*flow.Seal, error) {
+	seal, err := s.state.seals.ByBlockID(s.blockID)
+	if err != nil {
+		return nil, fmt.Errorf("could not look up latest seal: %w", err)
+	}
+	return seal, nil
+}
+
+// TODO inject results storage
+func (s *Snapshot) LatestResult() (*flow.ExecutionResult, error) {
+	seal, err := s.LatestSeal()
+	if err != nil {
+		return nil, fmt.Errorf("could not get latest seal: %w", err)
+	}
+	result, err := s.state.results.ByID(seal.ResultID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get latest result: %w", err)
+	}
+	return result, nil
+}
+
+func (s *Snapshot) SealingSegment() ([]*flow.Block, error) {
+	seal, err := s.LatestSeal()
+	if err != nil {
+		return nil, fmt.Errorf("could not get seal for sealing segment: %w", err)
+	}
+	head, err := s.state.blocks.ByID(s.blockID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get head block: %w", err)
+	}
+
+	// if this snapshot references the root block, the sealing segment
+	// consists only of the root block
+	segment := []*flow.Block{head}
+	if s.blockID == seal.BlockID {
+		return segment, nil
+	}
+
+	// for all other cases we walk through the chain backward until we reach
+	// the block referenced by the latest seal - the returned segment includes
+	// this block
+	nextID := head.Header.ParentID
+	for {
+		next, err := s.state.blocks.ByID(nextID)
+		if err != nil {
+			return nil, fmt.Errorf("could not get next block (id=%x): %w", nextID, err)
+		}
+
+		segment = append(segment, next)
+		if nextID == seal.BlockID {
+			break
+		}
+		nextID = next.Header.ParentID
+	}
```

Comment:
I've seen this pattern a few times. I think we could build an abstraction on top of the state, so that we don't need to always write this for-loop and create this intermediate value `nextID`:

```
err := traverseUnsealedFromBlock(head.Header.ParentID, func(header *flow.Header) {
  segment = append(segment, header)
})

```

Reply:
Great idea 👌

Reply:
Added here https://github.com/onflow/flow-go/pull/272/commits/f4e1578f27a8d3a38509009ac6cc0f1087bd69d9